src/dock.js

```diff
@@ -0,0 +1,536 @@
+'use strict'
+
+const _ = require('underscore-plus')
+const {CompositeDisposable} = require('event-kit')
+const PaneContainer = require('./pane-container')
+const TextEditor = require('./text-editor')
+
+const MINIMUM_SIZE = 100
+const DEFAULT_INITIAL_SIZE = 300
+const HANDLE_SIZE = 4
+const SHOULD_ANIMATE_CLASS = 'atom-dock-should-animate'
+const OPEN_CLASS = 'atom-dock-open'
+const RESIZE_HANDLE_RESIZABLE_CLASS = 'atom-dock-resize-handle-resizable'
+const TOGGLE_BUTTON_VISIBLE_CLASS = 'atom-dock-toggle-button-visible'
+const CURSOR_OVERLAY_VISIBLE_CLASS = 'atom-dock-cursor-overlay-visible'
+
+// Extended: A container at the edges of the editor window capable of holding items.
+// You should not create a `Dock` directly, instead use {Workspace::open}.
+module.exports = class Dock {
+  constructor (params) {
+    this.handleResizeHandleDragStart = this.handleResizeHandleDragStart.bind(this)
+    this.handleMouseMove = this.handleMouseMove.bind(this)
+    this.handleMouseUp = this.handleMouseUp.bind(this)
+    this.handleDrag = _.throttle(this.handleDrag.bind(this), 30)
+    this.handleDragEnd = this.handleDragEnd.bind(this)
+
+    this.location = params.location
+    this.widthOrHeight = getWidthOrHeight(this.location)
+    this.config = params.config
+    this.applicationDelegate = params.applicationDelegate
+    this.deserializerManager = params.deserializerManager
+    this.notificationManager = params.notificationManager
+    this.viewRegistry = params.viewRegistry
+
+    this.paneContainer = new PaneContainer({
+      config: this.config,
+      applicationDelegate: this.applicationDelegate,
+      deserializerManager: this.deserializerManager,
+      notificationManager: this.notificationManager,
+      views: this.viewRegistry
+    })
+
+    this.state = {
+      open: false,
+      shouldAnimate: false
+    }
+
+    this.subscriptions = new CompositeDisposable(
+      this.paneContainer.observePanes(pane => {
+        pane.onDidAddItem(this.handleDidAddPaneItem.bind(this))
+      }),
+      this.paneContainer.observePanes(pane => {
+        pane.onDidRemoveItem(this.handleDidRemovePaneItem.bind(this))
+      })
+    )
+  }
+
+  // FIXME(matthewwithanm: This is kinda gross. We need to get a view for the pane container so we
```

Comment:
I started switching over to that but I felt like it was a lot of change for this PR so I pulled back.  (Converting PaneContainer means converting Pane.)

But I did it now 😊 98ce09a